src/main/java/org/embulk/parser/json/JsonParserPlugin.java

```diff
@@ -167,40 +163,54 @@ public void run(TaskSource taskSource, Schema schema, FileInput input, PageOutpu
                 final String fileName = input.hintOfCurrentInputFileNameForLogging().orElse("-");
-
                 boolean evenOneJsonParsed = false;
-                try (JsonParser.Stream stream = newJsonStream(in, task)) {
-                    Value originalValue;
-                    while ((originalValue = stream.next()) != null) {
+                try (final JsonValueParser parser = newParser(in, task, parserBuilder)) {
+                    JsonValue originalValue;
+                    while ((originalValue = parser.readJsonValue()) != null) {
                         try {
-                            Value value = originalValue;
+                            JsonValue value = originalValue;
                             if (task.getRoot().isPresent()) {
                                 try {
-                                    value = jsonParser.parseWithOffsetInJsonPointer(originalValue.toJson(), task.getRoot().get());
+                                    value = JsonValueParser.builder(JSON_FACTORY)
+                                            .root(task.getRoot().get())
+                                            .build(originalValue.toJson())
+                                            .readJsonValue();
+                                    if (value == null) {
+                                        throw new JsonRecordValidateException("A Json record doesn't have given 'JSON pointer to root'.");
+                                    }
                                 } catch (JsonParseException e) {
+                                    /*
+                                     * When JsonParseException is thrown, it would be an error that
+                                     * the given JSON pointer doesn't match with the JSON object.
+                                     * We would return NULL when the pointer doesn't match, not throw Exception.
+                                     *
+                                     * NOTE: We may change the behavior.
+                                     * See: https://github.com/embulk/embulk/pull/1103#discussion_r255807991
+                                     */
                                     throw new JsonRecordValidateException("A Json record doesn't have given 'JSON pointer to root'.");
                                 }
                             }
 
-                            final Iterable<Value> recordValues;
+                            final Iterable<JsonValue> recordValues;
                             if (task.getFlattenJsonArray()) {
-                                if (!value.isArrayValue()) {
+                                if (!value.isJsonArray()) {
                                     throw new JsonRecordValidateException(
                                             String.format(
                                                     "A Json record must represent array value with '__experimental__flatten_json_array' option, but it's %s",
```

Comment:
Offtopic: `__experimental__flatten_json_array` -> `flatten_json_array`

Reply:
Thanks for your good catch! Exactly. Fixed it in: b8039b48869f43c58c1a6a3c9fab947751c76252